lib/HLSL/HLSignatureLower.cpp

```diff
@@ -1204,6 +1204,50 @@ void HLSignatureLower::GenerateDxilInputsOutputs(DXIL::SignatureKind SK) {
   }
 }
 
+bool HLSignatureLower::ValidateSemanticType(llvm::Function* F) {
+  bool result = true;
+  DxilFunctionAnnotation* funcAnnotation = HLM.GetFunctionAnnotation(F);
+  if (!funcAnnotation) {
+    return result;
+  }
+  for (Argument& arg : F->args()) {
+    DxilParameterAnnotation &paramAnnotation =
+      funcAnnotation->GetParameterAnnotation(arg.getArgNo());
+    llvm::StringRef semanticStr = paramAnnotation.GetSemanticString();
+    if (semanticStr.empty()) {
+      continue;
+    }
+    unsigned index;
+    StringRef baseSemanticStr;
+    Semantic::DecomposeNameAndIndex(semanticStr, &baseSemanticStr, &index);
+    const Semantic* semantic = Semantic::GetByName(baseSemanticStr);
+    Type* argTy = arg.getType();
+
+    if (argTy->isPointerTy())
+      argTy = cast<PointerType>(argTy)->getPointerElementType();
+
+    if (argTy->isArrayTy()) {
+      // Array type for arguments with specific qualifiers are expected.
+      // In this case, we do validation on array's element type.
+      DxilParamInputQual inputQual = paramAnnotation.GetParamInputQual();
+      if (inputQual == DxilParamInputQual::InputPatch ||
+        inputQual == DxilParamInputQual::InputPrimitive ||
+        inputQual == DxilParamInputQual::OutIndices ||
+        inputQual == DxilParamInputQual::OutPrimitives ||
+        inputQual == DxilParamInputQual::OutputPatch ||
+        inputQual == DxilParamInputQual::OutVertices) {
```

Comment:
Wish there was a better place to associate the input qualifier property with the expected array parameter shape (though this is a bit of an internal HL IR detail).  Even if it was in its own function somewhere.